include/atm_data_copy.h

```diff
@@ -0,0 +1,64 @@
+#undef __atm_data_src
+#ifdef _atm_src_ptr
+#  define __atm_data_src Atm(mygrid)%_atm_src_ptr
+#endif
+
+#undef __atm_data_dst
+#ifdef _atm_dst_ptr
+#  define __atm_data_dst _atm_dst_ptr
+#endif 
+
+#undef  __ik
+#undef  __beg_k_loop
+#undef  __end_k_loop
+#ifdef _atm_copy_3d
+#  define __ik ,k
+#  define __beg_k_loop do k=1,nk
+#  define __end_k_loop enddo
+#else
+#  define __ik
+#  define __beg_k_loop
+#  define __end_k_loop
+#endif
+
+#undef __lev
+#ifdef _atm_copy_lev
+#  define __lev ,:
+#else
+#  define __lev
+#endif
```

Comment:
I do not like all this ifdef/undef/define magic. This is very unreadable, difficult to debug and maintain. We should not use this kind of cpp string processing to try to write generic code. This should all be written in Fortran.

Reply:
@DusanJovic-NOAA I generally agree with you, but there are cases where writing the same (or almost the same) code fragment a large number of times in a single file increases the chances of introducing a bug. It also makes debugging and maintenance quite difficult since it would involve visually parsing almost identical code blocks.

The issue here is that we cannot define a Fortran function to replace each code block and improve modularity as well as readability. This case includes 74 almost identical code fragments and more may be added in the future. Using smaller building blocks that can be quickly updated and tested will improve resiliency for the entire code base.

Reply:
All you need is couple of fortran subroutines that accept 2d, 3d or 4d pointers, have a 2d, 3d or 4d do-loops and copy the values to the destination. I agree that we do not want to repeat all those 74 loops 74 times. But like I said using all those ifdefs/undefs/defines to achieve some kind of generality is not the way to go. I find this code extremely unreadable and difficult to reason about and debug.

Reply:
@DusanJovic-NOAA Could you please provide an example of such subroutine? I'd welcome a simpler solution.

The issue I see is that we can't pass each source array as an argument to a generic subroutine, since these arrays are components of a derived type array (_e.g._ `GFS_Data`) and we need to loop over each of its elements (number of grid blocks) to retrieve them.

Reply:
I believe atm_data_copy can be written as a simple subroutine, or few generic subroutines. gfs_data_copy, because of the GFS_data blocking, probably can not. But even so, I'd rather deal with (repeated) code like this:

```
!$omp parallel do default(shared) private(i,j,nb,ix)                                                                                 
      do j=jsc,jec                                                                                                                   
        do i=isc,iec                                                                                                                 
          nb = Atm_block%blkno(i,j)                                                                                                  
          ix = Atm_block%ixp(i,j)                                                                                                    
          datar82d(i-isc+1,j-jsc+1) = GFS_data(nb)%coupling%snow_cpl(ix) * rtimek                                                    
        enddo                                                                                                                        
      enddo                                                                                                                          
```
than something like:
```
          ! MEAN snow precipitation rate (kg/m2/s)
          case ('mean_fprec_rate')
#define _atm_src_ptr coupling%snow_cpl
#include "gfs_data_copy.h"
```
and hope that everyone understands that several lines above, somebody correctly defines:
```
#undef  _atm_src_cnv
#define _atm_src_cnv rtimek
```
so that scaling is done properly.
And what if somebody adds a new field after `mean_fprec_rate`, that does not have to be scaled, but forgets to `#undef __atm_src_cnv`. Constantly defining and un-defining all those cpp macros is in my opinion very error prone, much more than bunch of repeated loops, where at least we can see exactly what's going on. What if one needs more complicated scaling than simple multiplication factor. 

How would you debug (print) just one field. 

@junwang-noaa @climbfuji @DeniseWorthen What's your opinion on this?

Reply:
I agree that documentation needs to be added for the macros.